x/group/server/msg_server.go

```diff
@@ -0,0 +1,604 @@
+		ExecutorResult:      group.ProposalExecutorResultNotRun,
+		Timeout:             *endTime,
+		VoteState: group.Tally{
+			YesCount:     "0",
+			NoCount:      "0",
+			AbstainCount: "0",
+			VetoCount:    "0",
+		},
+	}
+	if err := m.SetMsgs(msgs); err != nil {
+		return nil, sdkerrors.Wrap(err, "create proposal")
+	}
+
+	id, err := s.proposalTable.Create(ctx, m)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "create proposal")
+	}
+
+	// TODO: add event?
+
+	return &group.MsgCreateProposalResponse{ProposalId: group.ProposalID(id)}, nil
+}
+
+func (s serverImpl) Vote(ctx types.Context, req *group.MsgVoteRequest) (*group.MsgVoteResponse, error) {
+	id := req.ProposalId
+	voters := req.Voters
+	choice := req.Choice
+	comment := req.Comment
+
+	if err := assertCommentSize(comment, s.maxCommentSize(ctx), "comment"); err != nil {
+		return nil, err
+	}
+	if len(voters) == 0 {
+		return nil, sdkerrors.Wrap(group.ErrEmpty, "voters")
+	}
+
+	blockTime, err := gogotypes.TimestampProto(ctx.BlockTime())
+	if err != nil {
+		return nil, err
+	}
+	proposal, err := s.getProposal(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	if proposal.Status != group.ProposalStatusSubmitted {
+		return nil, sdkerrors.Wrap(group.ErrInvalid, "proposal not open")
+	}
+	votingPeriodEnd, err := gogotypes.TimestampFromProto(&proposal.Timeout)
+	if err != nil {
+		return nil, err
+	}
+	if votingPeriodEnd.Before(ctx.BlockTime()) || votingPeriodEnd.Equal(ctx.BlockTime()) {
+		return nil, sdkerrors.Wrap(group.ErrExpired, "voting period has ended already")
+	}
+	var accountInfo group.GroupAccountInfo
+
+	address, err := sdk.AccAddressFromBech32(proposal.GroupAccount)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "group account")
+	}
+	if err := s.groupAccountTable.GetOne(ctx, address.Bytes(), &accountInfo); err != nil {
+		return nil, sdkerrors.Wrap(err, "load group account")
+	}
+	if proposal.GroupAccountVersion != accountInfo.Version {
+		return nil, sdkerrors.Wrap(group.ErrModified, "group account was modified")
+	}
+
+	electorate, err := s.getGroupInfo(ctx, accountInfo.GroupId)
+	if err != nil {
+		return nil, err
+	}
+	if electorate.Version != proposal.GroupVersion {
+		return nil, sdkerrors.Wrap(group.ErrModified, "group was modified")
+	}
+
+	// count and store votes
+	for _, voterAddr := range voters {
+		voter := group.GroupMember{GroupId: electorate.GroupId, Member: voterAddr}
+		if err := s.groupMemberTable.GetOne(ctx, voter.NaturalKey(), &voter); err != nil {
+			return nil, sdkerrors.Wrapf(err, "address: %s", voterAddr)
+		}
+		newVote := group.Vote{
+			ProposalId:  id,
+			Voter:       voterAddr,
+			Choice:      choice,
+			Comment:     comment,
+			SubmittedAt: *blockTime,
+		}
+		if err := proposal.VoteState.Add(newVote, voter.Weight); err != nil {
+			return nil, sdkerrors.Wrap(err, "add new vote")
+		}
+
+		if err := s.voteTable.Create(ctx, &newVote); err != nil {
+			return nil, sdkerrors.Wrap(err, "store vote")
+		}
```

Comment:
Should we also check if `voter` has `voted` already?

Reply:
this is already handled by the ORM which will return a `unique constraint violation` error if it already exists, there's also a test that covers this case but I'll add a comment to make it clearer.